### ISSUE TITLE
Xgboost training fails if input dataframe has vector type

### ISSUE THREAD

**Issue by viadea** [NVIDIA/spark-rapids-examples#108]
If the input dataframe has vector type inside, xgboost training fails with below error:

```
22/02/16 14:36:29 ERROR GpuXGBoostSpark: The job was aborted due to
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at ml.dmlc.xgboost4j.scala.spark.rapids.GpuUtils$.toColumnarRdd(GpuUtils.scala:49)
	at ml.dmlc.xgboost4j.scala.spark.rapids.GpuXGBoost$.trainOnGpuInternal(GpuXGBoost.scala:240)
	at ml.dmlc.xgboost4j.scala.spark.rapids.GpuXGBoost$.trainDistributedOnGpu(GpuXGBoost.scala:186)
	at ml.dmlc.xgboost4j.scala.spark.rapids.GpuXGBoost$.trainOnGpu(GpuXGBoost.scala:91)
	at ml.dmlc.xgboost4j.scala.spark.rapids.GpuXGBoost$.fitOnGpu(GpuXGBoost.scala:52)
	at ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier.fit(XGBoostClassifier.scala:170)
	at $line43.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:51)
	at $line43.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:56)
	at $line43.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:58)
	at $line43.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:60)
	at $line43.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:62)
	at $line43.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:64)
	at $line43.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:66)
	at $line43.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:68)
	at $line43.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:70)
	at $line43.$read$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:72)
	at $line43.$read$$iw$$iw$$iw$$iw$$iw.<init>(<console>:74)
	at $line43.$read$$iw$$iw$$iw$$iw.<init>(<console>:76)
	at $line43.$read$$iw$$iw$$iw.<init>(<console>:78)
	at $line43.$read$$iw$$iw.<init>(<console>:80)
	at $line43.$read$$iw.<init>(<console>:82)
	at $line43.$read.<init>(<console>:84)
	at $line43.$read$.<init>(<console>:88)
	at $line43.$read$.<clinit>(<console>)
	at $line43.$eval$.$print$lzycompute(<console>:7)
	at $line43.$eval$.$print(<console>:6)
	at $line43.$eval.$print(<console>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.call(IMain.scala:745)
	at scala.tools.nsc.interpreter.IMain$Request.loadAndRun(IMain.scala:1021)
	at scala.tools.nsc.interpreter.IMain.$anonfun$interpret$1(IMain.scala:574)
	at scala.reflect.internal.util.ScalaClassLoader.asContext(ScalaClassLoader.scala:41)
	at scala.reflect.internal.util.ScalaClassLoader.asContext$(ScalaClassLoader.scala:37)
	at scala.reflect.internal.util.AbstractFileClassLoader.asContext(AbstractFileClassLoader.scala:41)
	at scala.tools.nsc.interpreter.IMain.loadAndRunReq$1(IMain.scala:573)
	at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:600)
	at sun.reflect.GeneratedMethodAccessor42.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.nvidia.spark.rapids.ColumnarRdd$.convert(ColumnarRdd.scala:52)
	at com.nvidia.spark.rapids.ColumnarRdd.convert(ColumnarRdd.scala)
	... 53 more
Caused by: java.lang.IllegalArgumentException: Cannot convert [label: float, feature: float ... 1 more field] to GPU columnar Set(org.apache.spark.mllib.linalg.VectorUDT@f71b0bce) are not currently supported data types for columnar.
	at org.apache.spark.sql.rapids.execution.InternalColumnarRddConverter$.extractRDDColumnarBatch(InternalColumnarRddConverter.scala:665)
	at org.apache.spark.sql.rapids.execution.InternalColumnarRddConverter$.convert(InternalColumnarRddConverter.scala:718)
	at org.apache.spark.sql.rapids.execution.InternalColumnarRddConverter.convert(InternalColumnarRddConverter.scala)
	... 59 more
```

Below is a minimum reproduce notebook in scala
```
import org.apache.spark.sql.SparkSession
sc.stop()

// Build the spark session and data reader as usual
val spark = SparkSession.builder.appName("xgboost_vector_test").getOrCreate
val reader = spark.read

import ml.dmlc.xgboost4j.scala.spark.{XGBoostClassifier, XGBoostClassificationModel}
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator
import org.apache.spark.sql.types.{FloatType, IntegerType, StructField, StructType}
import org.apache.spark.sql.types._
import org.apache.spark.sql.Row

val trainPath = "/home/xxx/data/xgboost_vector_test"

// with Vector
val rows = spark.sparkContext.parallelize(
  List(
    Row(0.0, 1.2, org.apache.spark.mllib.linalg.Vectors.dense(1.0, 2.0))
  )
)

val schema = List(
  StructField("label", DoubleType, true),
  StructField("feature", DoubleType, true),
  StructField("a_vector", new org.apache.spark.mllib.linalg.VectorUDT, true)
)

val df = spark.createDataFrame(
  rows,
  StructType(schema)
)

df.show()
df.printSchema
df.write.format("parquet").mode("overwrite").save(trainPath)

val trainSet = reader.parquet(trainPath)
trainSet.printSchema

val labelColName = "label"
val featureNames = Array("feature")

val commParamMap = Map(
  "eta" -> 0.1,
  "gamma" -> 0.1,
  "missing" -> 0.0,
  "max_depth" -> 10,
  "max_leaves" -> 256,
  "objective" -> "binary:logistic",
  "grow_policy" -> "depthwise",
  "min_child_weight" -> 30,
  "lambda" -> 1,
  "scale_pos_weight" -> 2,
  "subsample" -> 1,
  "nthread" -> 1,
  "num_round" -> 100)

val xgbParamFinal = commParamMap ++ Map("tree_method" -> "gpu_hist", "num_workers" -> 1)

val xgbClassifier = new XGBoostClassifier(xgbParamFinal)
      .setLabelCol(labelColName)
      // === diff ===
      .setFeaturesCols(featureNames)

xgbClassifier.fit(trainSet)
```

**Test env:**
Standalone Spark cluster
Spark 3.1.1
22.02 snapshot rapids-spark and cudf jars
xgboost4j_3.0-1.4.2-0.2.0.jar
xgboost4j-spark_3.0-1.4.2-0.2.0.jar

This is a customer blocker issue.

**Comment by viadea** [comment]
I know vector is not supported by Spark RAPIDS plugin as of now. 
Does it mean our xgboost jar can only work with supported data types?

**Comment by raford** [comment]
A question in addition to the one from @viadea that deals with the same issue.

Assuming that covariates in vector format are not supported, what is the best way to process heterogeneous data in the feature space for model training using only PySpark? 

That is, given covariates that can be either numeric or categorical, what, if any, preprocessing needs to be performed with the categorical features to ensure that a XGBoost model can be trained via the GPU implementation, and do you have a working example of this?

The feature space in the example notebooks is all numeric, but dealing with categorical variables is very important.

Thank you,


Ray

**Comment by GaryShen2008** [comment]
@wbo4958 Can you provide what data type is supported by XGBoost?
As I remember, only numeric data type can be supported. 

To encode the categorical variables to numeric ones, we have one example code for Criteo dataset [here](https://github.com/NVIDIA/DeepLearningExamples/blob/master/TensorFlow2/Recommendation/DLRM/preproc/spark_data_utils.py).

**Comment by wbo4958** [comment]
XGBoost itself supports numeric types for features, labels, weights, margins.

For this issue, it is SparkRapids not supporting Vector Type. The operators like file scan will not run on GPU because of unsupported Vector Type, So when converting the Dataframe to RDD[Table], SparkRapids throws an exception.

The workaround is to drop the columns with the vector type.
